source/_posts/2018-02-26-URLComponents.md

```diff
@@ -0,0 +1,367 @@
+---
+title:  URLComponents
+date: 2018-02-26 10:00:00
+tags: ios, urlcomponents
+authorIds:
+- pebo
+categories:
+- iOS
+---
+When writing mobile apps, more often than not we have to communicate with a backend to get some data to present to the user.
+
+The shared language between frontend and backend is usually URLs combined with either a POST body or maybe some URL query parameters, depending on what the backend offers.
+
+In this post we will look at [URLComponents](https://developer.apple.com/documentation/foundation/urlcomponents), a small but very useful helper that will become your new best friend when composing or decomposing URLs (we promise!)
+
+But before we look at `URLComponents`, lets first look at a world without `URLComponents` just to make you extra grateful that such a thing exists.
+
+## Setting the Stage
+OK, lets assume that we need to use a backend service for searching and showing details about TV shows.
+
+### Searching
+ Our hard working backend developing friends have made a service for us to use. The base URL looks like this:
+
+`https://showsknownfrom.tv/search`
+
+(no...it doesn't exist)
+
+And it offers the following query parameters:
+
+- q: A query string we would like to search for
+- order (optional parameter): Do we want the search results ascending or descending
+- number (optional parameter): Max number of search results to return
+
+### Showing Details
+To see details about a TV show, we get a URL to a HTML page from the backend which we then present in a WebView.
+
+The URL looks like this:
+
+`https://showsknownfrom.tv/12345678`
+
+And can have an optional URL parameter called `featured`. If featured is present and is true, we would like to show a flashing red `UILabel` with the caption "FEATURED" over our `WebView` (note: this feature has not been cleared with the designer yet!)
+
+## Composing and Decomposing URLs - The Hard Way
+### Composing URLs for Searching
+The stage is set and we are ready to look at how to call our search service in our pre-URLComponents world.
+
+To generate a `URL` in our iOS app, we could write some code like this:
+
+```swift
+func searchURL(with q: String, optionalParameters: [String: String]? = nil) -> URL? {
+    var urlString = "https://showsknownfrom.tv/search"
+
+    //Append query
+    urlString.append("?q=\(q)")
```

Comment:
You could point out that URLComponents handle the case of not having query parameter nicely by not adding the `?` when there are none.